libsplinter/src/peer/component.rs

```diff
@@ -13,3 +13,211 @@
+        );
+
+        // validate that the services are in the directory
+        self.state
+            .get_service(service_id)
+            .map_err(|err| ServiceForwardingError::InternalError {
+                context: format!("unable to look up service {}", service_id),
+                source: Some(Box::new(err)),
+            })?
+            .ok_or_else(|| ServiceForwardingError::SenderNotRegistered)?;
+
+        let recipient_service = self
+            .state
+            .get_service(&recipient_id)
+            .map_err(|err| ServiceForwardingError::InternalError {
+                context: format!("unable to look up service {}", &recipient_id),
+                source: Some(Box::new(err)),
+            })?
+            .ok_or_else(|| ServiceForwardingError::RecipientNotRegistered)?;
+
+        if self.splinter_node.id() != recipient_service.node().id() {
+            // Construct a circuit direct message, and send it to the service peer:
+            let (circuit_name, recipient_id) = recipient_id.into_parts();
+            let mut direct_msg = CircuitDirectMessage::new();
+            direct_msg.set_circuit(circuit_name);
+            direct_msg.set_sender(service_id.service_id().to_string());
+            direct_msg.set_recipient(recipient_id);
+            direct_msg.set_payload(service_msg.payload);
+            if let Some(correlation_id) = service_msg.correlation_id {
+                direct_msg.set_correlation_id(correlation_id);
+            }
+
+            let bytes = direct_msg.write_to_bytes().map_err(|err| {
+                ServiceForwardingError::InternalError {
+                    context: "unable to serialize circuit direct message envelope".into(),
+                    source: Some(Box::new(err)),
+                }
+            })?;
+
+            let msg = create_message(bytes, CircuitMessageType::CIRCUIT_DIRECT_MESSAGE).map_err(
+                |err| ServiceForwardingError::InternalError {
+                    context: "unable to create circuit/network envelope".into(),
+                    source: Some(Box::new(err)),
+                },
+            )?;
+
+            self.network_sender
+                .send(recipient_service.node().id().into(), msg)
+                .map(|_| ForwardResult::Sent)
+                .map_err(|_| ServiceForwardingError::InternalError {
+                    context: format!(
+                        "unable to send admin direct message to {}",
+                        recipient_service.node().id()
+                    ),
+                    source: None,
+                })
+        } else if let Some(component_id) = recipient_service.peer_id() {
+            Ok(ForwardResult::LocalReReroute(
+                component_id.into(),
+                service_msg,
+            ))
+        } else {
+            Err(ServiceForwardingError::InternalError {
+                context: format!(
+                    "Service {} was registered without a component id",
```

Comment:
a/id/ID/, also the capitalization of the first word is not consistent with the other errors in this module.